src/VisualStudio/CSharp/Impl/LanguageService/CSharpCodeCleanupFixer.cs

```diff
@@ -320,37 +340,138 @@ private async Task<bool> FixTextBufferAsync(TextBufferCodeCleanUpScope textBuffe
+            {
+                cancellationToken = cancellationTokenSource.Token;
+
+                var buffer = textBufferScope.SubjectBuffer;
+                using (var scope = context.OperationContext.AddScope(allowCancellation: true, description: EditorFeaturesResources.Applying_changes))
+                {
+                    var progressTracker = new ProgressTracker((description, completed, total) =>
+                    {
+                        if (scope != null)
+                        {
+                            scope.Description = description;
+                            scope.Progress.Report(new ProgressInfo(completed, total));
+                        }
+                    });
+
+                    var document = buffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                    var newDoc = await FixDocumentAsync(document, context.EnabledFixIds, progressTracker, cancellationToken).ConfigureAwait(true);
 
                     await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                    cancellationToken.ThrowIfCancellationRequested();
+
                     return document.Project.Solution.Workspace.TryApplyChanges(newDoc.Project.Solution, progressTracker);
                 }
             }
         }
+
+        private async Task<Solution> FixSolutionAsync(Solution solution, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            progressTracker.AddItems(solution.ProjectIds.Count);
+            foreach (var projectId in solution.ProjectIds)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var project = solution.GetProject(projectId);
+                progressTracker.Description = project.Name;
+                var newProject = await FixProjectAsync(project, enabledFixIds, new ProgressTracker(), cancellationToken).ConfigureAwait(false);
+                solution = newProject.Solution;
+                progressTracker.ItemCompleted();
+            }
+
+            return solution;
+        }
+
+        private async Task<Project> FixProjectAsync(Project project, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            if (project.LanguageServices.GetService<ICodeCleanupService>() is null)
+            {
+                return project;
+            }
+
+            progressTracker.AddItems(project.DocumentIds.Count);
+            foreach (var documentId in project.DocumentIds)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var document = project.GetDocument(documentId);
+                progressTracker.Description = document.Name;
+                var fixedDocument = await FixDocumentAsync(document, enabledFixIds, new ProgressTracker(), cancellationToken).ConfigureAwait(false);
```

Comment:
can we keep pass in the progress tracker? rather than pass in new one to discard any document level changes if possible?

Reply:
The discarding was intentional. We don't have a good way to handle nested progress. I added comments explaining it.